File: 한양대학교/컴퓨터네트워크/Assignment2/Kraken.py
import time
import urllib.parse
import hashlib
import hmac
import base64
import requests

class Kraken(object):
    api_url = "https://api.kraken.com"

    def get_kraken_signature(self, urlpath, data):
        postdata = urllib.parse.urlencode(data)
        encoded = (str(data['nonce']) + postdata).encode()
        message = urlpath.encode() + hashlib.sha256(encoded).digest()

        mac = hmac.new(base64.b64decode(self.api_sec), message, hashlib.sha512)
        sigdigest = base64.b64encode(mac.digest())
        return sigdigest.decode()

    def kraken_request(self, uri_path, data):
        headers = {}
        headers['API-Key'] = self.api_key
        headers['API-Sign'] = self.get_kraken_signature(uri_path, data)
        req = requests.post((self.api_url + uri_path), headers=headers, data=data)
        print(f"응답 코드: {req.status_code}")
        return req

    def place_order(self, volume, pair, buy_price, order_type="buy"):
        resp = self.kraken_request('/0/private/AddOrder', {
            "nonce": str(int(1000 * time.time())),
            "ordertype": "limit",
            "type": order_type,
            "volume": volume,
            "pair": pair,
            "price": buy_price
        })
        return resp.json()

    def get_balance(self):
        resp = self.kraken_request('/0/private/Balance', {
            "nonce": str(int(1000 * time.time()))
        })
        return resp.json()


if __name__ == '__main__':
    kr = Kraken()
    response1 = kr.get_balance()
    response2 = kr.place_order(volume="0.01", pair="XBTUSD", buy_price="40000", order_type="sell")
    print(response1)
    print(response2)

File: 한양대학교/컴퓨터네트워크/Assignment2/Kraken.rb
require 'net/http'
require 'uri'
require 'openssl'
require 'base64'
require 'time'
require 'json'

class Kraken
  API_URL = "https://api.kraken.com"

  def get_kraken_signature(urlpath, data)
    postdata = URI.encode_www_form(data)
    nonce = data['nonce']
    encoded = nonce + postdata

    digest = OpenSSL::Digest::SHA256.new
    message = urlpath + digest.digest(encoded)

    mac = OpenSSL::HMAC.digest(OpenSSL::Digest.new('sha512'), Base64.decode64(API_SEC), message)
    Base64.strict_encode64(mac)
  end

  def kraken_request(uri_path, data)
    url = URI.parse(API_URL + uri_path)
    http = Net::HTTP.new(url.host, url.port)
    http.use_ssl = true

    request = Net::HTTP::Post.new(url)
    request['API-Key'] = API_KEY
    request['API-Sign'] = get_kraken_signature(uri_path, data)
    request.set_form_data(data)

    response = http.request(request)
    puts response.code
    JSON.parse(response.body)
  end

  def place_order(volume, pair, buy_price, order_type = "buy")
    kraken_request('/0/private/AddOrder', {
      "nonce" => (Time.now.to_f * 1000).to_i.to_s,
      "ordertype" => "limit",
      "type" => order_type,
      "volume" => volume,
      "pair" => pair,
      "price" => buy_price
    })
  end

  def get_balance
    kraken_request('/0/private/Balance', {
      "nonce" => (Time.now.to_f * 1000).to_i.to_s
    })
  end
end

if __FILE__ == $0
  kr = Kraken.new
  response1 = kr.get_balance
  response2 = kr.place_order("0.01", "XBTUSD", "40000", "sell")

  puts response1
  puts response2
end
